Documentation/ApiOverview/Context/Index.rst

```diff
@@ -18,13 +18,27 @@ track of, for example, the current time, if a user is logged in and which
 workspace is currently accessed.
 
 The :php:`\TYPO3\CMS\Core\Context\Context` object can be retrieved via
-:ref:`dependency injection <DependencyInjection>`:
+:ref:`dependency injection <DependencyInjection>` in the injection method:
+
+..  literalinclude:: _MyControllerUsingDependencyInjection.php
+    :language: php
+    :caption: EXT:my_extension/Classes/Controller/MyController.php
+
+This requires under some circumstances also an entry in the :file:`Services.yaml` file (:ref:`more in <ServicesYaml>`):
```

Comment:
Which circumstances? Additionally, why start with the non-recommended variant? constructor injection should be preferred.

Reply:
No, read the DOCS and you will see that injection methods are recommended!

Reply:
Actually https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/DependencyInjection/Index.html#constructor-injection says:

```
The recommended usage is [constructor injection](https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/DependencyInjection/Index.html#constructor-injection)
```

Maybe you are referring to outdated documentation or for older TYPO3 versions? We should then try to correct these places if you could point us towards it, because injection methods nowadays are really not the recommendation.

Reply:
It is the most current documentation which clearly says that constructor injections should not be used!

https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/DependencyInjection/Index.html#method-injection



Now assume the abstract class is provided by TYPO3 core and the consuming class is provided by an extension. If the abstract class would use constructor injection, the consuming class would need to know the dependencies of the abstract, add its own dependency to the constructor, and then call parent::__construct($logger) to satisfy the dependency of the abstract. This would hardcode all dependencies of the abstract into extending classes. If later the abstract is changed and another dependency is added to the constructor, this would break consuming classes since they did not know that.

Differently put: When core classes "pollute" __construct() with dependencies, the core can not add dependencies without being breaking. This is the reason why for example the extbase AbstractController uses inject*() methods for its dependencies: Extending classes can then use constructor injection, do not need to call parent::__construct(), and the core is free to change dependencies of the abstract.

In general, when the core provides abstract classes that are expected to be extended by extensions, the abstract class should use inject*() methods instead of constructor injection. Extensions of course can follow this idea in similar scenarios.

This construct has some further implications: Abstract classes should think about making their dependency properties private, so extending classes can not rely on them. Furthermore, classes that should not be extended by extensions are free to use constructor injection and should be marked final, making sure they can't be extended to allow internal changes.

Reply:
The inject methods should be used if you provide an abstract class or a class which should be extended from, as the extending class cannot use constructor injection that way - when the super class changes the injections via constructor that will break also the extending class.

Using inject methods is an edge case. When you have a custom controller/middleware/CLI command/... (which will mostly neved be extended) you can and should use the constructor injection.

Reply:
There are still many other drawbacks for a constructor based DI. If a constructor already has many parameters then you have difficulties to discern the order of the parameters. If you add another parameter, then all makeInstance calls must be modified. This leads to more effort than simply adding another inject method.

Reply:
Constructor injection makes GU::makeInstance() calls mostly superfluous.

Reply:
What? How should I get the object without a makeInstance call?

Reply:
Via Dependency Injection. That is the whole idea of that

Reply:
Yes, the passed object is provided via dependency injection. But the class which receives this object must be created by MakeInstance. 
As far as I understand only the TYPO3 Core classes are injected. Or does also a TYPO3 Extension inject its own classes into its base controller classes?

Reply:
sure, all you need is a services.yaml. or services.php. This is well documented